File: Puzzles/2019-11/solutions/november_puzzle.py
def is_prime(n):
    if n == 0 or n == 1:
        return False
    count = 2
    while count < n:
        if n % count == 0:
            return False
        else:
            count += 1
    return True

def is_leap_year(n):
    if n % 400 == 0:
        return True
    else:
        if n % 100 == 0:
            return False
        elif n % 4 == 0:
            return True
        else:
            return False


year = 2
date_year = 2
month = 11
day = 333
special_days = 0
weekday = 6
november_count = 29
thursday_count = 4
while year < 3000:
    if is_prime(date_year):
        if month == 11:
            if thursday_count == 4:
                if weekday == 6:
                    if is_prime(november_count):
                        special_days += 1
                        thursday_count = 0
                        month += 1
                        november_count = 0
                    else:
                        thursday_count = 0
                
    day += 1
    weekday += 1
    if weekday == 8:
        weekday = 1
    if is_leap_year(year):
        if day == 306:
            month = 11
        if day == 335:
            month += 1
            november_count = 0
            thursday_count = 0
    else:
        if day == 305:
            month = 11
        if day == 334:
            month += 1
            november_count = 0
            thursday_count = 0
    if month == 11 and weekday == 5:
        thursday_count += 1
    if month == 11:
        november_count += 1
    if is_leap_year(year):
        if day == 367:
            year += 1
            date_year += 1
            day = 1
    else:
        if day == 366:
            year += 1
            date_year += 1
            day = 1
    if date_year == 101:
        date_year = 1
    
print(special_days)
    
    
    

    

File: Puzzles/2019-11/solutions/novemberpuzzle.java
public class novemberpuzzle {

    /**
     * returns the date of black friday
     */
    public static int getDate(int year){
        int[] dates = {24,23,29,28,27,26,25};
        int tracker = 0;
        for (int i =1; i <= year; i++){
            
            if (i % 400 == 0){
                tracker+=2;
            }
            else if (i % 4 == 0 && i % 100 !=0){
                tracker+=2;
            }
            else{
                tracker++;
            }
        }
        return dates[tracker % 7];
    }


    public static int numBlackFriday(){

        int count = 0;                     // counts num of black friday primes
        int[] calender = new int[2];       // {day , year}

        for (int i = 1; i < 3000; i++){    // cycles through every year
            calender[0] = getDate(i);       // c[0] = day
            calender[1] = i % 100;         // c[1] =  last 2 digits of year

            boolean isPrime = true;

            /*
            ===================
            * i = year , j = items in calender , k = every num < (j / 2)
            ===================
            */
            for (int j: calender){         // for elem in cal
                while (isPrime){
                    if (j == 0 || j == 1){ 
                        isPrime = false;
                        break;
                    }
                    for (int k = 2; k < (j/2) + 1; k++){ //prime checker
                        
                        if (j % k == 0){ // if j(nums in calender array) is divisble by any number(K); not prime
                            isPrime = false;
                            break;
                        }
                    }
                    break;
                }
                
            }
            if (isPrime){ // if isPrime is still true after all the checks
                count++;

                isPrime = false; //reset isPrime
            }
        }
        
        return count;
    }

    public static void main(String[] args){
        System.out.println(numBlackFriday());

    }


}